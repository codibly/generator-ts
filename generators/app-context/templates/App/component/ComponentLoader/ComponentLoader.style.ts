import styled from '@emotion/styled';
import Typography, { TypographyProps } from '@material-ui/core/Typography';
import { ComponentType } from 'react';

export const LoaderWrapper = styled.div({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '100%',
  height: '100%',
  padding: 15,
  boxSizing: 'border-box'
});

export const LoaderMessage = styled(Typography)({
  display: 'inline-block',
  padding: '1em'
}) as ComponentType<TypographyProps>;
