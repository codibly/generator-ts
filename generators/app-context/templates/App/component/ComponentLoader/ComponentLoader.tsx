import CircularProgress from '@material-ui/core/CircularProgress';
import * as React from 'react';
import { FunctionComponent } from 'react';
import { LoadingComponentProps } from 'react-loadable';
import { Centered } from '../Centered/Centered';
import { ErrorFallback } from '../ErrorFallback/ErrorFallback';
import { ErrorFallbackAction } from '../ErrorFallback/ErrorFallbackAction/ErrorFallbackAction';
import { ErrorFallbackContent } from '../ErrorFallback/ErrorFallbackContent/ErrorFallbackContent';
import { ErrorFallbackTitle } from '../ErrorFallback/ErrorFallbackTitle/ErrorFallbackTitle';
import { LoaderMessage } from './ComponentLoader.style';

export namespace ComponentLoader {
  export type Props = LoadingComponentProps;
}

export const ComponentLoader: FunctionComponent<ComponentLoader.Props> = ({
  timedOut,
  pastDelay,
  error,
  retry
}) => {
  if (error) {
    return (
      <ErrorFallback>
        <ErrorFallbackTitle>Loading error</ErrorFallbackTitle>
        <ErrorFallbackContent>
          An error occurred during the application loading and this view could not be displayed.
          Please, try to <ErrorFallbackAction onClick={retry}>reload this view</ErrorFallbackAction>
        </ErrorFallbackContent>
      </ErrorFallback>
    );
  } else if (timedOut) {
    return (
      <Centered>
        <CircularProgress color="secondary" />
        <LoaderMessage variant="body1">Taking a long time...</LoaderMessage>
      </Centered>
    );
  } else if (pastDelay) {
    return (
      <Centered>
        <CircularProgress />
      </Centered>
    );
  } else {
    return null;
  }
};
