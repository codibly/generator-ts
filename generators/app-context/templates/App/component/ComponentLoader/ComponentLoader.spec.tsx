import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ComponentLoader } from './ComponentLoader';

describe('<ComponentLoader />', () => {
  it('should render nothing before delay pass', () => {
    const { container } = render(
      <ComponentLoader
        pastDelay={false}
        isLoading={true}
        timedOut={false}
        error={false}
        retry={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render loader after delay pass', () => {
    const { container } = render(
      <ComponentLoader
        pastDelay={true}
        isLoading={true}
        timedOut={false}
        error={false}
        retry={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render taking too long after timeout pass', () => {
    const { container, getByText } = render(
      <ComponentLoader
        pastDelay={true}
        isLoading={true}
        timedOut={true}
        error={false}
        retry={jest.fn()}
      />
    );

    expect(getByText('Taking a long time...')).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render error page and retry button on error', () => {
    const retry = jest.fn();
    const { container, getByText } = render(
      <ComponentLoader
        pastDelay={false}
        isLoading={false}
        timedOut={false}
        error={true}
        retry={retry}
      />
    );

    expect(
      getByText(
        /An error occurred during the application loading and this view could not be displayed/
      )
    ).toBeDefined();
    fireEvent.click(getByText('reload this view'));

    expect(retry).toHaveBeenCalled();
    expect(container.firstChild).toMatchSnapshot();
  });
});
