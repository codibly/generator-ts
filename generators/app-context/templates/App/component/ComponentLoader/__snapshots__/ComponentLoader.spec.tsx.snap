// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComponentLoader /> should render error page and retry button on error 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  text-align: center;
}

.emotion-0 {
  color: #74A1BE;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

<section
  class="emotion-1"
>
  <h6
    class="MuiTypography-root MuiTypography-h6"
  >
    Loading error
  </h6>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2"
  >
    An error occurred during the application loading and this view could not be displayed. Please, try to 
    <span
      class="emotion-0"
    >
      reload this view
    </span>
  </h6>
</section>
`;

exports[`<ComponentLoader /> should render loader after delay pass 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  class="emotion-0"
>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;

exports[`<ComponentLoader /> should render nothing before delay pass 1`] = `null`;

exports[`<ComponentLoader /> should render taking too long after timeout pass 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  display: inline-block;
  padding: 1em;
}

<div
  class="emotion-1"
>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
  <p
    class="MuiTypography-root emotion-0 MuiTypography-body1"
  >
    Taking a long time...
  </p>
</div>
`;
